<a href="https://colab.research.google.com/github/Hannah1123/Pytorch_note/blob/main/pytorchbasic.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import torch
import torch.nn.functional as fn

In [ ]:
a = torch.eye(3)
b = torch.eye(3)[None, :, :]
print('a', a)
print('size', a.shape)
print('b', b)
print('size', b.shape)


a tensor([[1., 0., 0.],
        [0., 1., 0.],
        [0., 0., 1.]])
size torch.Size([3, 3])
b tensor([[[1., 0., 0.],
         [0., 1., 0.],
         [0., 0., 1.]]])
size torch.Size([1, 3, 3])


In [ ]:
[None, :, :]

tensor([[[1., 0., 0.],
         [0., 1., 0.],
         [0., 0., 1.]]])

In [ ]:
x = torch.tensor([1, 2, 3])
x0 = x.expand(2, 3)
print(x0)
print(x.shape)
print(x0.shape)

tensor([[1, 2, 3],
        [1, 2, 3]])
torch.Size([3])
torch.Size([2, 3])


In [ ]:
x1 = torch.randn(2, 1, 1, 4)
print(x1)

tensor([[[[-0.1782, -0.6373, -1.1717,  0.2201]]],


        [[[ 1.1863, -0.0233,  0.4874, -0.5745]]]])


In [ ]:
x2 = x1.expand(-1, 2, 3, -1)
print(x2)
print(x2.shape)

tensor([[[[-0.1782, -0.6373, -1.1717,  0.2201],
          [-0.1782, -0.6373, -1.1717,  0.2201],
          [-0.1782, -0.6373, -1.1717,  0.2201]],

         [[-0.1782, -0.6373, -1.1717,  0.2201],
          [-0.1782, -0.6373, -1.1717,  0.2201],
          [-0.1782, -0.6373, -1.1717,  0.2201]]],


        [[[ 1.1863, -0.0233,  0.4874, -0.5745],
          [ 1.1863, -0.0233,  0.4874, -0.5745],
          [ 1.1863, -0.0233,  0.4874, -0.5745]],

         [[ 1.1863, -0.0233,  0.4874, -0.5745],
          [ 1.1863, -0.0233,  0.4874, -0.5745],
          [ 1.1863, -0.0233,  0.4874, -0.5745]]]])
torch.Size([2, 2, 3, 4])


##repeat()

repeat()是size乘以一个倍数

In [ ]:
x = torch.tensor([1, 2, 3])
x3 = x.repeat(2, 3)
print(x3)
print(x.shape)
print(x3.shape)

tensor([[1, 2, 3, 1, 2, 3, 1, 2, 3],
        [1, 2, 3, 1, 2, 3, 1, 2, 3]])
torch.Size([3])
torch.Size([2, 9])


In [ ]:
x3 = torch.randn(2, 3, 4)
x3.repeat(3, 1, 3).shape

torch.Size([6, 3, 12])

##simple indices

In [ ]:
a = torch.arange(5)
print('a：', a)

a： tensor([0, 1, 2, 3, 4])


In [ ]:
a.view(5, 1, 1)

tensor([[[0]],

        [[1]],

        [[2]],

        [[3]],

        [[4]]])

In [ ]:
a.shape

torch.Size([5])

In [ ]:
batch = a.view(5, 1, 1).repeat(1, 10, 3)
print(batch)

tensor([[[0, 0, 0],
         [0, 0, 0],
         [0, 0, 0],
         [0, 0, 0],
         [0, 0, 0],
         [0, 0, 0],
         [0, 0, 0],
         [0, 0, 0],
         [0, 0, 0],
         [0, 0, 0]],

        [[1, 1, 1],
         [1, 1, 1],
         [1, 1, 1],
         [1, 1, 1],
         [1, 1, 1],
         [1, 1, 1],
         [1, 1, 1],
         [1, 1, 1],
         [1, 1, 1],
         [1, 1, 1]],

        [[2, 2, 2],
         [2, 2, 2],
         [2, 2, 2],
         [2, 2, 2],
         [2, 2, 2],
         [2, 2, 2],
         [2, 2, 2],
         [2, 2, 2],
         [2, 2, 2],
         [2, 2, 2]],

        [[3, 3, 3],
         [3, 3, 3],
         [3, 3, 3],
         [3, 3, 3],
         [3, 3, 3],
         [3, 3, 3],
         [3, 3, 3],
         [3, 3, 3],
         [3, 3, 3],
         [3, 3, 3]],

        [[4, 4, 4],
         [4, 4, 4],
         [4, 4, 4],
         [4, 4, 4],
         [4, 4, 4],
         [4, 4, 4],
         [4, 4, 4],
         [4, 4, 4],
         [4, 4, 4],
         [4,

In [ ]:
idx = torch.randint(0, 5, (3,)).repeat(5, 10, 1)
print(idx.shape)

In [ ]:
points = torch.rand(5, 5, 10)


In [ ]:
new_points = points[batch, idx, :]


In [ ]:
new_points.shape

torch.Size([5, 10, 3, 10])

##indices

###longtensor  
x和mask形状不必相同

In [ ]:
#x, mask都为1-D
x1 = torch.Tensor([5,4,3,2,1])

mask1 = torch.LongTensor([1,0,2])
x1[mask1]

tensor([4., 5., 3.])

In [ ]:
#x 2-D
#mask 1-D


x2 = torch.Tensor([[ 0.,  1.,  2.,  3.,  4.],
           [ 5.,  6.,  7.,  8.,  9.],
           [10., 11., 12., 13., 14.]])

mask1 = torch.LongTensor([0,1,2])
 
x2[mask1]

tensor([[ 0.,  1.,  2.,  3.,  4.],
        [ 5.,  6.,  7.,  8.,  9.],
        [10., 11., 12., 13., 14.]])

In [ ]:
#不论x是几维的，mask为LongTensor时都是按照mask中的顺序排列x[mask[i]]

mask2 = torch.LongTensor([0,2,1])
 
x2[mask2]

tensor([[ 0.,  1.,  2.,  3.,  4.],
        [10., 11., 12., 13., 14.],
        [ 5.,  6.,  7.,  8.,  9.]])

In [ ]:
#x 2-D
#mask 2-D


x2 = torch.Tensor([[ 0.,  1.,  2.,  3.,  4.],
           [ 5.,  6.,  7.,  8.,  9.],
           [10., 11., 12., 13., 14.]])
 
mask1 = torch.LongTensor([[0,1,2],[0,1,2]])
 
print(x2[mask1])
 
print('x2.shape: ',x2.shape)
print('mask1.shape: ',mask1.shape)
print('x2[mask1].shape: ', x2[mask1].shape)

tensor([[[ 0.,  1.,  2.,  3.,  4.],
         [ 5.,  6.,  7.,  8.,  9.],
         [10., 11., 12., 13., 14.]],

        [[ 0.,  1.,  2.,  3.,  4.],
         [ 5.,  6.,  7.,  8.,  9.],
         [10., 11., 12., 13., 14.]]])
x2.shape:  torch.Size([3, 5])
mask1.shape:  torch.Size([2, 3])
x2[mask1].shape:  torch.Size([2, 3, 5])


In [ ]:
#我们将mask中的数字i当作x在第一个纬度上的索引就好了

mask2 = torch.LongTensor([[0,0,0],[0,0,0]])
 
x2[mask2]


tensor([[[0., 1., 2., 3., 4.],
         [0., 1., 2., 3., 4.],
         [0., 1., 2., 3., 4.]],

        [[0., 1., 2., 3., 4.],
         [0., 1., 2., 3., 4.],
         [0., 1., 2., 3., 4.]]])

In [ ]:
x2[[0,0,0],[0,0,0]]  #什么区别？？？

tensor([0., 0., 0.])

In [ ]:
#x 3-D
#mask 3-D

x = torch.arange(0,27).view(3,3,3)
print(x)
print(x[[1,2],[1,2],[2,0]])  # x[1,1,2]和x[2,2,0]
print(x[[2,1,0],[0],[0]])  # x[2,0,0]和x[1,0,0]和x[0,0,0]


tensor([[[ 0,  1,  2],
         [ 3,  4,  5],
         [ 6,  7,  8]],

        [[ 9, 10, 11],
         [12, 13, 14],
         [15, 16, 17]],

        [[18, 19, 20],
         [21, 22, 23],
         [24, 25, 26]]])
tensor([14, 24])
tensor([18,  9,  0])
ERROR! Session/line number was not unique in database. History logging moved to new session 59


In [ ]:
print(x[[2,1,0],[0],:])

tensor([[18, 19, 20],
        [ 9, 10, 11],
        [ 0,  1,  2]])


##sort


*   input (Tensor) – the input tensor  
形式上与 numpy.narray 类似  
*   dim (int, optional) – the dimension to sort along  
维度，对于二维数据：dim=0 按列排序，dim=1 按行排序，默认 dim=1

*   descending (bool, optional) – controls the sorting order (ascending or descending)  
降序，descending=True 从大到小排序，**descending=False 从小到大排序，默认 descending=Flase**






  


会返回原来的indices

In [ ]:
x = torch.randn(3,4)
print(x)

tensor([[-0.4366, -0.1434,  2.4892, -0.2917],
        [ 0.8097, -1.4708, -0.1354,  0.8627],
        [ 0.9184, -0.7001,  0.9967, -1.1782]])


In [ ]:
sorted, indices = torch.sort(x)  #按行从小到大排序
print('sorted:', sorted)
print('indices: ', indices)

sorted: tensor([[-0.4366, -0.2917, -0.1434,  2.4892],
        [-1.4708, -0.1354,  0.8097,  0.8627],
        [-1.1782, -0.7001,  0.9184,  0.9967]])
indices:  tensor([[0, 3, 1, 2],
        [1, 2, 0, 3],
        [3, 1, 0, 2]])


In [ ]:
sorted1, indices1 = torch.sort(x, dim=0)  #按列从小到大排序
print('sorted1:', sorted1)
print('indices1: ', indices1)

sorted1: tensor([[-0.4366, -1.4708, -0.1354, -1.1782],
        [ 0.8097, -0.7001,  0.9967, -0.2917],
        [ 0.9184, -0.1434,  2.4892,  0.8627]])
indices1:  tensor([[0, 1, 1, 2],
        [1, 2, 2, 0],
        [2, 0, 0, 1]])


##max

形式： torch.max(input, dim, keepdim=False, out=None) -> (Tensor, LongTensor)  
按维度dim 返回最大值，并且返回索引。

In [ ]:
a = torch.randn(3,4)
print(a)

tensor([[-0.4535, -0.9024, -0.6280, -0.4881],
        [ 0.2952,  0.2242, -1.0195,  0.7486],
        [ 0.2327,  1.5419, -0.3626, -0.5007]])


In [ ]:
maxa0 = torch.max(a, dim=0)[0]
print('dim=0时max为： ', maxa0)

maxa1 = torch.max(a, dim=1)[0]
print('dim=1时max为： ', maxa1)

dim=0时max为：  tensor([ 0.2952,  1.5419, -0.3626,  0.7486])
dim=1时max为：  tensor([-0.4535,  0.7486,  1.5419])


In [ ]:
b = torch.randn(2,3,4)
print(b)

tensor([[[ 1.2064,  0.1783, -0.7743,  0.0690],
         [ 0.0592,  1.9157,  0.7327,  0.5757],
         [-0.1002, -0.1501,  0.9726, -0.0315]],

        [[ 0.4632,  0.4688, -0.9933, -0.3292],
         [ 0.9207,  0.3172, -0.6343, -0.8197],
         [ 0.6468,  0.5792, -1.6884,  0.6103]]])


size为[2， 3， 4]时，  
dim=0时， 含有 2 个元素，2个元素内取大{1.2064>0.4632}  
dim=1时， 每一维含有 3 个元素，3个元素内取大{1.2064>0.0592>-0.1002}  
dim=2时， b含有 4 个元素， 4个元素内取大{1.2064>0.1783>0.0690>-0.7743}

In [ ]:
maxb0 = torch.max(b, dim=0)[0]
print('dim=0时max为： ', maxb0)

maxb1 = torch.max(b, dim=1)[0]
print('dim=1时max为： ', maxb1)

maxb2 = torch.max(b, dim=2)[0]
print('dim=2时max为： ', maxb2)

dim=0时max为：  tensor([[ 1.2064,  0.4688, -0.7743,  0.0690],
        [ 0.9207,  1.9157,  0.7327,  0.5757],
        [ 0.6468,  0.5792,  0.9726,  0.6103]])
dim=1时max为：  tensor([[ 1.2064,  1.9157,  0.9726,  0.5757],
        [ 0.9207,  0.5792, -0.6343,  0.6103]])
dim=2时max为：  tensor([[1.2064, 1.9157, 0.9726],
        [0.4688, 0.9207, 0.6468]])


In [ ]:
maxb0 = torch.max(b, dim=0)[0]
print('dim=0时max为： ', maxb0.shape)

maxb1 = torch.max(b, dim=1)[0]
print('dim=1时max为： ', maxb1.shape)

maxb2 = torch.max(b, dim=2)[0]
print('dim=2时max为： ', maxb2.shape)

dim=0时max为：  torch.Size([3, 4])
dim=1时max为：  torch.Size([2, 4])
dim=2时max为：  torch.Size([2, 3])


##softmax

##log_sofmax

##sum

##cat

torch.cat(seq, dim=0, out=None) → Tensor

在指定的维度dim上对序列seq进行连接操作。

参数：

seq (sequence of Tensors) - Python序列或相同类型的张量序列  
dim (int, optional) - 沿着此维度连接张量  
out (Tensor, optional) - 输出参数

In [ ]:
a0=torch.Tensor([[[[1,1,1,1],[2,2,2,2]]]])
a1=torch.Tensor([[[[3,3,3,3],[4,4,4,4]]]])

In [ ]:
print(a0.shape)

torch.Size([1, 1, 2, 4])


In [ ]:
cat0 = torch.cat((a0,a1),dim=0)
print(cat0)

tensor([[[[1., 1., 1., 1.],
          [2., 2., 2., 2.]]],


        [[[3., 3., 3., 3.],
          [4., 4., 4., 4.]]]])


In [ ]:
cat0.shape

torch.Size([2, 1, 2, 4])

In [ ]:
cat2 = torch.cat((a0,a1),dim=2)
print(cat2)

tensor([[[[1., 1., 1., 1.],
          [2., 2., 2., 2.],
          [3., 3., 3., 3.],
          [4., 4., 4., 4.]]]])


In [ ]:
cat2.shape

torch.Size([1, 1, 4, 4])

In [ ]:
cat3 = torch.cat((a0,a1),dim=3)
print(cat3)

tensor([[[[1., 1., 1., 1., 3., 3., 3., 3.],
          [2., 2., 2., 2., 4., 4., 4., 4.]]]])


torch.rand()

torch.randn()

torch.normal()

torch.linespace()

##矩阵相乘

In [ ]:
a = torch.ones(2, 3, 4)
print('a: ', a)
b = torch.Tensor([1, 2, 3])
print('b: ', b)

a:  tensor([[[1., 1., 1., 1.],
         [1., 1., 1., 1.],
         [1., 1., 1., 1.]],

        [[1., 1., 1., 1.],
         [1., 1., 1., 1.],
         [1., 1., 1., 1.]]])
b:  tensor([1., 2., 3.])


In [ ]:
b = b.repeat(2,1)
print('b: ', b)

b:  tensor([[1., 2., 3.],
        [1., 2., 3.]])


In [ ]:
b.shape

torch.Size([2, 3])

In [ ]:
b= b.view(2, 3, 1)

In [ ]:
print('b: ', b)

b:  tensor([[[1.],
         [2.],
         [3.]],

        [[1.],
         [2.],
         [3.]]])


In [ ]:
mul = a * b
print('output:', mul)

output: tensor([[[1., 1., 1., 1.],
         [2., 2., 2., 2.],
         [3., 3., 3., 3.]],

        [[1., 1., 1., 1.],
         [2., 2., 2., 2.],
         [3., 3., 3., 3.]]])


In [ ]:
mul.shape

torch.Size([2, 3, 4])